File: my_robot_bringup/launch/turtlesim_catch_all_turtles/turtlesim_catch_all_turtles/turtle_controller.py
#!/usr/bin/python3

import rclpy
from math import sqrt,atan2,pi
from rclpy.node import Node
from turtlesim.msg import Pose
from functools import partial
from geometry_msgs.msg import Twist
from my_robot_interfaces.msg import Turtle
from my_robot_interfaces.msg import Turtles
from my_robot_interfaces.srv import KillTurtle

'''
    Functionality:
    - Run a control loop to reach a given target point. 
      The first turtle on the screen "turtle1" will be the "master" turtle control. 
    - The control loop will a P controller
    - Subscribe to the /alive_turtles topic to get all current turtles with coordinates.
      From that info, select a turtle
    - When a turtle has been caught by the master turtle, call the service /catch_turtle advertised by
      the turtle_spawner node

'''

class turtleControllerNode(Node): 
    def __init__(self):
        super().__init__("turtle_controller") 
        
        #Control parameters
        self.T = 0.01
        self.P = 3
        self.P_ang = 6
        self.I = 0
        self.D = 0

        #Publishers, subscribers and services
        self.timer_controller_ = self.create_timer(self.T,self.control)
        self.subscriber_pose_  = self.create_subscription(Pose,"/turtle1/pose",self.callback_get_current_pos,10)
        self.alive_turtles_    = self.create_subscription(Turtles,"/alive_turtles",self.getTurtle,10)
        self.publisher_cmdVel_ = self.create_publisher(Twist,"/turtle1/cmd_vel",10)
        self.service_client_   = self.create_client(KillTurtle,"/catch_turtle")
        
        #Parameters
        self.declare_parameter("catch_closest_turtle_first",True)
        self.closest_turtle_ = self.get_parameter("catch_closest_turtle_first").value
        self.use_sim_time = self.get_parameter("use_sim_time").value


        #Other control parameters
        self.current_Pos_X_ = 0
        self.current_Pos_Y_ = 0
        self.current_Pos_angTheta_ = 0

        self.catchingATurtle_ = False
        self.turtleList_ = []

        self.turtleNameToCatch_ = ""
        self.setPointX_ = 0
        self.setPointY_ = 0
        self.setPointTheta_ = 0


        self.difX = 0.0
        self.difY = 0.0
        self.target_ang = 0.0
        self.target_ang_raw = 0.0
        self.eTheta = 0.0
        self.eDist = 0.0
        self.error_anterior_ = 0

        self.threshold_linear_ = 0.05
        self.threshold_angular_ = 0.25
        self.threshold_angular_2_ = 10

        self.turtlesKilled_ = []

        #For showing PIDs variables in cmd
        self.flag_debug = True
        
        self.current_time_ = self.get_current_time()

        self.get_logger().info("turtleControllerNode is up! ...")


    def control(self):
        #Control loop

        if self.catchingATurtle_ == False:
            #If I'm not catching a turtle -> get a turtle to catch
            

            #Choose if catch by closest turtle or by spawing order
            if self.closest_turtle_ == True:
                turtle = self.get_closest_turtle()
            else:
                turtle = self.get_first_turtle()

            #If a turtle is founded -> Get data from turtle
            if turtle.name != "":

                self.turtleNameToCatch_ = turtle.name
                self.setPointX_ = turtle.x
                self.setPointY_ = turtle.y
                self.setPointZ_ = turtle.theta

                self.catchingATurtle_ = True
                
            else:

                self.setPointX_ = 0.0
                self.setPointY_ = 0.0
                self.setPointZ_ = 0.0
                self.difX = 0.0
                self.difY = 0.0
                self.eDist = 0.0
                self.eTheta = 0.0

        else:
            #If I'm catching a turtle -> continue doing control to catch a turtle

            self.catchingATurtle_ = True

            #Control algorithm
            self.control_algorithm2()

            #For visualizing info
            if self.flag_debug:
                if abs(self.eTheta)  > abs(self.error_anterior_):
                    self.get_logger().error("--------------------------------------------------------")
                self.get_logger().info("Error dist: " + str(self.eDist))
                self.get_logger().info("difX: " + str(self.difX))
                self.get_logger().info("difY: " + str(self.difY))
                self.get_logger().info("target_ang: " + str(self.target_ang))
                self.get_logger().info("current_Pos_angTheta_: " + str(self.current_Pos_angTheta_))
                self.get_logger().info("Error theta: " + str(self.eTheta))
                self.error_anterior_ = self.eTheta





    def control_algorithm(self):

        self.get_error_signals()

        #Make msg for command velocity
        msg = Twist()

        if abs(self.eDist) < self.threshold_linear_ and abs(self.eTheta) < self.threshold_angular_:
            #If I'm close enough -> turtle catched

            self.stop_and_kill(msg)

        else:
            
            if abs(self.eDist) > self.threshold_linear_: 
                msg.linear.x = self.P  * self.eDist
            else:
                msg.linear.x = 0.0

            if abs(self.eTheta) > self.threshold_angular_: 
                msg.angular.z = self.P_ang * self.eTheta
            else: 
                msg.angular.z = 0.0


        self.publisher_cmdVel_.publish(msg)




    def control_algorithm2(self):

            self.get_error_signals()

            #Make msg for command velocity
            msg = Twist()

            if abs(self.eDist) < self.threshold_linear_ and abs(self.eTheta) < self.threshold_angular_:
                #If I'm close enough -> turtle catched

                self.stop_and_kill(msg)

            else:
                
                

                if abs(self.eTheta) > self.threshold_angular_2_: 
                    msg.linear.x = 0.0
                    msg.angular.z = self.P_ang * self.eTheta
                else: 
                    if abs(self.eTheta) > self.threshold_angular_: 
                        msg.angular.z = self.P_ang * self.eTheta
                        if abs(self.eDist) > self.threshold_linear_: 
                            msg.linear.x = self.P  * self.eDist
                        else:
                            msg.linear.x = 0.0
                    else:
                        if abs(self.eDist) > self.threshold_linear_: 
                            msg.linear.x = self.P  * self.eDist
                        else:
                            msg.linear.x = 0.0
                        msg.angular.z = 0.0

            self.publisher_cmdVel_.publish(msg)


    def get_error_signals(self):

        self.difX = self.setPointX_ - self.current_Pos_X_
        self.difY = self.setPointY_ - self.current_Pos_Y_

        self.eDist  = sqrt((self.difX)**2 + (self.difY)**2)
        self.target_ang = atan2(self.difY,self.difX)

        self.eTheta =  self.target_ang - self.current_Pos_angTheta_

        if self.eTheta > pi:
            self.eTheta -= 2*pi
        elif self.eTheta < -pi:
            self.target_ang += 2*pi
        
        
        

    def stop_and_kill(self,msg):
        #If I'm close enough -> turtle catched

        #Creating the catch message
        msg_turtle_kill = KillTurtle.Request()
        turtle_catched = Turtle()
        turtle_catched.name = str(self.turtleNameToCatch_)
        turtle_catched.x = self.setPointX_
        turtle_catched.y = self.setPointY_
        turtle_catched.theta = float(self.setPointTheta_)

        msg_turtle_kill.turtle = turtle_catched.name

        #Sending string Turtle name
        future = self.service_client_ .call_async(msg_turtle_kill)
        future.add_done_callback(partial(self.callback_turtle_killed,turtle_name=turtle_catched.name))

        #Tracking turtles catched
        self.turtlesKilled_.append(turtle_catched.name)

        #Stopping turtle
        msg.linear.x = 0.0
        msg.angular.z = 0.0

        self.catchingATurtle_ = False





    def get_current_time(self):
        secs, nsecs = self.get_clock().now().seconds_nanoseconds()
        return secs + nsecs/1000000000.0 

    def callback_get_current_pos(self,msg):
        self.current_Pos_X_ = msg.x
        self.current_Pos_Y_ = msg.y
        self.current_Pos_angTheta_ = msg.theta


    def getTurtle(self,msg):
        self.turtleList_ = msg.turtles

    def get_first_turtle(self):
        #Returns the first turtle founded

        turtleToCatch = Turtle()
        for turtle in self.turtleList_:
            if turtle.name not in self.turtlesKilled_:
                turtleToCatch = turtle
        return turtleToCatch

    def get_closest_turtle(self):
        #Returns the closest turtle founded

        minimumDistance = 99999
        turtleToCatch = Turtle()
        for turtle in self.turtleList_:
            if turtle.name not in self.turtlesKilled_:
                distance = sqrt((self.current_Pos_X_ - turtle.x)**2 + (self.current_Pos_Y_ - turtle.y)**2)
                if distance < minimumDistance:
                    minimumDistance = distance
                    turtleToCatch = turtle
    
        self.get_logger().info("Turtle to catch: " + str(turtleToCatch.name))

        return turtleToCatch

    def callback_turtle_killed(self,future,turtle_name):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info("Turtle catched: " + str(turtle_name))
                self.turtleList_ = []
            else:
                self.get_logger().warn("Turtle NOT catched: " + str(turtle_name))
        except:
            self.get_logger().error("Turtle NOT catched. Error in service!")


 
 
def main(args=None):
    rclpy.init(args=args)
    node = turtleControllerNode() 
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: turtlesim_catch_all_turtle_cpp/src/turtle_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/turtle.hpp"
#include "my_robot_interfaces/msg/turtles.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "my_robot_interfaces/srv/kill_turtle.hpp"
#include "turtlesim/msg/pose.hpp"

#include <cmath>

/*

    Functionality

    Run a control loop (for example using a timer with a high rate) to reach a given target point. The first turtle on the screen “turtle1” will be the “master” turtle to control. To control the turtle you can subscribe to /turtle1/pose and publish to /turtle1/cmd_vel.

    The control loop will use a simplified P controller.

    Subscribe to the /alive_turtles topic to get all current turtles with coordinates. From that info, select a turtle to target (to catch).

    When a turtle has been caught by the master turtle, call the service /catch_turtle advertised by the turtle_spawner node.

    Parameters:

    catch_closest_turtle_first

    use_sim_time

*/

class turtleControllerNode : public rclcpp::Node
{
public:
    turtleControllerNode() : Node("turtle_controller")
    {


        
        this->declare_parameter("catch_closest_turtle_first", true);

        catch_closest_turtle_first_ = this->get_parameter("catch_closest_turtle_first").as_bool();

        subscriber_alive_turtles_ = this->create_subscription<my_robot_interfaces::msg::Turtles>("/alive_turtles",
                                                                                                 10,
                                                                                                 std::bind(&turtleControllerNode::callback_alive_turtles,
                                                                                                           this,
                                                                                                           std::placeholders::_1));

        subscriber_turtle_pose_ = this->create_subscription<turtlesim::msg::Pose>("/turtle1/pose",
                                                                                  10,
                                                                                  std::bind(&turtleControllerNode::callback_get_pose,
                                                                                            this, 
                                                                                            std::placeholders::_1));


        publisher_cmd_vel_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel",10);

        service_client_catch_turtle_ = this->create_client<my_robot_interfaces::srv::KillTurtle>("/catch_turtle");

        control_loop_timer_ = this->create_wall_timer(std::chrono::milliseconds((int)T*1000),std::bind(&turtleControllerNode::control_loop,this));

        catching_a_turtle_ = false;


        RCLCPP_INFO(this->get_logger(),"turtle_controller is up!");

    }

    void control_loop()
    {

        if (catching_a_turtle_ == true)
        {
            control_algorithm();
        }
        else
        {
            publish_velocity(0.0,0.0);

            try
            {
                turtle_to_catch_ = get_turtle();
                if (turtle_to_catch_.name != "")
                {
                    catching_a_turtle_ = true;

                    set_point_X_ = turtle_to_catch_.x;
                    set_point_Y_ = turtle_to_catch_.y;
                    
                    ePos = 0.0;
                    eTheta = 0.0;

                    difX = 0.0;
                    difY = 0.0;

                    RCLCPP_INFO(this->get_logger(),"Turtle to catch: %s", turtle_to_catch_.name.c_str());
                }
                else
                {
                    RCLCPP_WARN(this->get_logger(),"No turtle available!");
                }
            }
            catch(const std::exception& e)
            {
                set_point_X_ = 0.0;
                set_point_Y_ = 0.0;
                ePos = 0.0;
                eTheta = 0.0;
                difX = 0.0;
                difY = 0.0;
                turtle_to_catch_.name = "";
                RCLCPP_ERROR(this->get_logger(),"No turtle available! - Error");
            }
            
        }

    }

    void control_algorithm()
    {
        difX = set_point_X_ - posX_;
        difY = set_point_Y_ - posY_;

        ePos = sqrt(pow(difX,2) + pow(difY,2));
        steering_angle = (float) atan2(difY,difX);
        eTheta =  steering_angle - theta_;

        if (eTheta > M_PI)
            eTheta -= 2*M_PI;
        else
        {
            if (eTheta < -M_PI)
                eTheta += 2*M_PI;
        }

        if (abs(eTheta) > eTheta_threshold2)
        {
            publish_velocity(0,P_ang*eTheta);

        }

        
        else
        {
            
            if (abs(ePos) > abs(ePos_threshold))
            {
                publish_velocity(P*ePos,P_ang*eTheta);

            }
            else
            {

                //Turtle catched!
                publish_velocity(0.0,0.0);
                pop_turtle_catched();
                
                RCLCPP_INFO(this->get_logger(),"Turtle catched: %s", turtle_to_catch_.name.c_str());
                catching_a_turtle_ = false;
                kill_turtle_threads_.push_back(std::make_shared<std::thread>(std::bind(&turtleControllerNode::kill_turtle,this,turtle_to_catch_.name)));

            }
            

        }
        
    }


    void kill_turtle(std::string turtle_name)
    {
        
        while(!service_client_catch_turtle_->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(),"Waiting for server");
        }

        auto request_kill_turtle_ = std::make_shared<my_robot_interfaces::srv::KillTurtle::Request>();

        request_kill_turtle_->turtle = turtle_name;

        auto future = service_client_catch_turtle_->async_send_request(request_kill_turtle_);

        try
        {
            auto response = future.get();

            if (response->success == true)
                RCLCPP_INFO(this->get_logger(),"Turtle catched!. Turtle: %s",turtle_name.c_str());
            else
                RCLCPP_ERROR(this->get_logger(),"Turtle NOT catched!. Turtle: %s",turtle_name.c_str());

        }
        catch(const std::exception& e)
        {
            RCLCPP_ERROR(this->get_logger(),"Error in calling /catch_turtle service");
        }
        


    }

    void callback_alive_turtles(const my_robot_interfaces::msg::Turtles::SharedPtr msg)
    {
        alive_turtles_vector_ = msg->turtles;
    }

    my_robot_interfaces::msg::Turtle get_turtle()
    {
        my_robot_interfaces::msg::Turtle turtle;
        float min_distance = 9999.0;
        float current_distance = 9999.0;
        int pos_turtle_founded = 0;

        if ((int)alive_turtles_vector_.size() > 0)
        // if there is a turtle to catch
        {
            turtle = alive_turtles_vector_.at(0);
            

            if (catch_closest_turtle_first_ == true)
            {
                // search and return closest turtle
                for (int i = 1; i < (int)alive_turtles_vector_.size(); i++)
                {
                    
                    current_distance = sqrt(pow(alive_turtles_vector_.at(i).x - posX_,2)+pow(alive_turtles_vector_.at(i).y - posY_,2));
                    if (current_distance < min_distance)
                    {
                        min_distance = current_distance;
                        turtle = alive_turtles_vector_.at(i);
                        pos_turtle_founded = i;
                    }
                }

                
            }
            else
            {
                // return the first turtle appeared and not catched
                turtle = alive_turtles_vector_.at(0);
                pos_turtle_founded = 0;

            }

            
            if ((int)catched_turtles_vector_.size() > 0)
            {
                for (int i = 1; i < (int)catched_turtles_vector_.size(); i++)
                    if (catched_turtles_vector_.at(i).name == turtle.name)
                    {
                        alive_turtles_vector_.erase(alive_turtles_vector_.begin()+ pos_turtle_founded);
                        turtle.name = "";
                        break;
                    }
            }
            

            return turtle;

        }


        //if there isn't any turtle, I use try-catch block where this function is called 
        return turtle;

    }

    void pop_turtle_catched()
    {
        my_robot_interfaces::msg::Turtle turtle;

        if ((int)alive_turtles_vector_.size() > 0)
        // if there is a turtle to catch
        {
            for (int i = 1; i < (int)alive_turtles_vector_.size(); i++)
            {
                turtle = alive_turtles_vector_.at(i);
                if (turtle.name == turtle_to_catch_.name)
                {
                    alive_turtles_vector_.erase(alive_turtles_vector_.begin()+ i);
                    catched_turtles_vector_.push_back(turtle);
                    break;
                } 
            }
        }
    }

    void callback_get_pose(const turtlesim::msg::Pose::SharedPtr msg)
    {
        posX_ = msg->x;
        posY_ = msg->y;
        theta_ = msg->theta;
    }

    void publish_velocity(float linearVel, float angularVel)
    {
        geometry_msgs::msg::Twist cmd_vel_ = geometry_msgs::msg::Twist();

        cmd_vel_.linear.x = linearVel;
        cmd_vel_.angular.z = angularVel;

        publisher_cmd_vel_->publish(cmd_vel_);

    }

private:
    bool catch_closest_turtle_first_;
    bool catching_a_turtle_;

    float posX_ = 0.0;
    float posY_ = 0.0;
    float theta_ = 0.0;

    float set_point_X_ = 0.0;
    float set_point_Y_ = 0.0;

    float ePos = 0.0;
    float eTheta = 0.0;

    float difX = 0.0;
    float difY = 0.0;
    float steering_angle = 0.0;

    //PID parameters
    float T = 0.01;
    float P = 2;
    float P_ang = 6;
    float I = 0.1;
    float D = 0.1;

    float ePos_threshold = 0.1;
    float eTheta_threshold = 0.1;
    float eTheta_threshold2 = 0.785;

    my_robot_interfaces::msg::Turtle turtle_to_catch_ = my_robot_interfaces::msg::Turtle();
    rclcpp::Client<my_robot_interfaces::srv::KillTurtle>::SharedPtr service_client_catch_turtle_;
    rclcpp::Subscription<my_robot_interfaces::msg::Turtles>::SharedPtr subscriber_alive_turtles_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscriber_turtle_pose_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_cmd_vel_;
    std::vector<my_robot_interfaces::msg::Turtle> alive_turtles_vector_;
    std::vector<my_robot_interfaces::msg::Turtle> catched_turtles_vector_;
    rclcpp::TimerBase::SharedPtr control_loop_timer_;
    std::vector<std::shared_ptr<std::thread>> kill_turtle_threads_;
    
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<turtleControllerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}